source/common/upstream/upstream_impl.cc

```diff
@@ -276,132 +203,72 @@ UpstreamLocalAddressSelectorImpl::buildBaseSocketOptions(
-        std::make_shared<Network::ConnectionSocket::Options>();
-
-    Network::Socket::appendOptions(additional_upstream_local_address.socket_options_,
-                                   base_socket_options);
-    Network::Socket::appendOptions(additional_upstream_local_address.socket_options_,
-                                   cluster_socket_options);
+Envoy::Upstream::UpstreamLocalAddressSelectorConstSharedPtr createUpstreamLocalAddressSelector(
+    const envoy::config::cluster::v3::Cluster& cluster_config,
+    const absl::optional<envoy::config::core::v3::BindConfig>& bootstrap_bind_config) {
 
-    upstream_local_addresses_.push_back(additional_upstream_local_address);
+  UpstreamLocalAddressSelectorFactory* local_address_selector_factory;
+  // The ``upstream_bind_config`` field on Cluster overrides the ``bootstrap_bind_config``
+  // entirely. Thus we ignore the ``local_address_selector`` on the bootstrap bind config if
+  // ``upstream_bind_config`` is specified without a ``local_address_selector``.
+  const envoy::config::core::v3::TypedExtensionConfig* const local_address_selector_config =
+      cluster_config.has_upstream_bind_config()
+          ? (cluster_config.upstream_bind_config().has_local_address_selector()
+                 ? &cluster_config.upstream_bind_config().local_address_selector()
+                 : nullptr)
+          : (bootstrap_bind_config.has_value() &&
+                     bootstrap_bind_config->has_local_address_selector()
+                 ? &bootstrap_bind_config->local_address_selector()
+                 : nullptr);
+  if (local_address_selector_config) {
+    local_address_selector_factory =
+        Config::Utility::getAndCheckFactory<UpstreamLocalAddressSelectorFactory>(
+            *local_address_selector_config, false);
+  } else {
+    envoy::extensions::upstream::local_address_selector::v3::DefaultLocalAddressSelector
+        default_config;
+    envoy::config::core::v3::TypedExtensionConfig typed_extension;
+    typed_extension.mutable_typed_config()->PackFrom(default_config);
+    local_address_selector_factory =
+        Config::Utility::getAndCheckFactory<UpstreamLocalAddressSelectorFactory>(typed_extension,
+                                                                                 false);
+  }
+  OptRef<const envoy::config::core::v3::BindConfig> bind_config_to_use;
+  if (cluster_config.has_upstream_bind_config()) {
+    bind_config_to_use.emplace(cluster_config.upstream_bind_config());
+  } else if (bootstrap_bind_config.has_value()) {
+    bind_config_to_use.emplace(*bootstrap_bind_config);
   }
+  return local_address_selector_factory->createLocalAddressSelector(
+      bind_config_to_use,
+      buildBaseSocketOptions(cluster_config,
+                             bootstrap_bind_config.value_or(envoy::config::core::v3::BindConfig{})),
+      buildClusterSocketOptions(
+          cluster_config, bootstrap_bind_config.value_or(envoy::config::core::v3::BindConfig{})),
```

Comment:
We should be able to combine the base and cluster socket options before pass to the selector

Reply:
I actually implemented it that way and then realised that the socket options in `extra_source_addresses` overrides only the [`cluster_socket_options`](https://github.com/envoyproxy/envoy/blob/main/source/common/upstream/upstream_impl.cc#L371)

Reply:
Emm... that makes me concerned that if we expose those socket options to the address selector extension, then those extensions may customize the way of merging those socket options. That isn't part of our API definition, we didn't say those selectors will change the socket options merging behavior.


One way to resolve that I'm thinking is we keep all those bind config parsing in the Envoy core code. Then we add two interfaces for the selector extensions, one is for validating those addresses (for default ext, ensure there is one address for each version), another one is for selecting the address. And this two interfaces, only accept the addresses as parameter.

Reply:
My major concern is this. @adisuissa   WDYT about this?

Reply:
@soulxu I've removed the ability for the extension to apply socket options from the bind config. PTAL.

Reply:
Oops, sorry, i will review again

Reply:
No problem! I should have responded to this comment after the commit.